README.md
# JuicyPixels-scale-dct

> Scale [JuicyPixels](https://github.com/Twinside/Juicy.Pixels) images with DCT

[![Build Status](https://travis-ci.org/phadej/JuicyPixels-scale-dct.svg?branch=master)](https://travis-ci.org/phadej/JuicyPixels-scale-dct)
[![Hackage](https://img.shields.io/hackage/v/JuicyPixels-scale-dct.svg)](http://hackage.haskell.org/package/JuicyPixels-scale-dct)
[![Stackage LTS 2](http://stackage.org/package/JuicyPixels-scale-dct/badge/lts-2)](http://stackage.org/lts-2/package/JuicyPixels-scale-dct)
[![Stackage LTS 3](http://stackage.org/package/JuicyPixels-scale-dct/badge/lts-3)](http://stackage.org/lts-3/package/JuicyPixels-scale-dct)
[![Stackage Nightly](http://stackage.org/package/JuicyPixels-scale-dct/badge/nightly)](http://stackage.org/nightly/package/JuicyPixels-scale-dct)

[[JuicyPixels](https://github.com/Twinside/Juicy.Pixels) is a Haskell library
to load &amp; save pictures, but it doesn't provide any image manipulation
functionality.

---

## Example image

![phadej](https://raw.githubusercontent.com/phadej/JuicyPixels-scale-dct/master/phadej.png)

### smaller

![phadej](https://raw.githubusercontent.com/phadej/JuicyPixels-scale-dct/master/phadej-small.png)

### very small

![phadej](https://raw.githubusercontent.com/phadej/JuicyPixels-scale-dct/master/phadej-smaller.png)

### larger

![phadej](https://raw.githubusercontent.com/phadej/JuicyPixels-scale-dct/master/phadej-large.png)


CHANGELOG.md
# 0.1.2

- Add `scaleWithKernel` to convolute the image. Usaful for applyng unsharp mask!

# 0.1.1.2

- Make test suite pass, even if `cat.jpg` is absent

# 0.1.1.1

- Remove test dependency on `JuicyPixels-util`

# 0.1.1.0

- Bundle missing test file


src/Codec/Picture/ScaleDCT.hs
{-# LANGUAGE FlexibleContexts    #-}
{-# LANGUAGE RankNTypes          #-}
{-# LANGUAGE ScopedTypeVariables #-}
-----------------------------------------------------------------------------
-- |
-- Module      :  Code.Picture.ScaleDCT
-- Copyright   :  (C) 2015 Oleg Grenrus
-- License     :  BSD3
-- Maintainer  :  Oleg Grenrus <oleg.grenrus@iki.fi>
--
-- Scale pictures using Discrete Cosine Transform.
--
module Codec.Picture.ScaleDCT (scale, scaleWithKernel) where

import Prelude ()
import Prelude.Compat

import Codec.Picture
       (Image (..), PixelRGBA8 (..), Traversal, generateImage, imagePixels)
import Control.Applicative (Const (..))
import Data.Array.CArray
       (CArray, amap, array, bounds, elems, listArray, size, (!))
import Data.Coerce         (Coercible, coerce)
import Data.Ix             (inRange, range)
import Data.Monoid         (Endo (..))
import Data.Word           (Word8)
import Math.FFT            (dct1N, dct2N, dct3N)

type Array2D = CArray (Int, Int) Double

-- | Scale the image using DCT transform.
scale
    :: (Int, Int)        -- ^ Output width, height
    -> Image PixelRGBA8  -- ^ Input image
    -> Image PixelRGBA8  -- ^ Output image
scale dim = scaleImpl (cut dim)

-- | Scale the image using DCT transform.
--
-- Convolute /result/ image with a symmetric kernel.
-- See <https://en.wikipedia.org/wiki/Symmetric_convolution>
--
-- Identity kernel:
--
-- @
-- k 0 0 = 1
-- k _ _ = 0
-- @
--
-- Sharpen:
--
-- @
-- k 0 0 = 1.75
-- k 0 1 = -0.125
-- k 1 0 = -0.125
-- k 1 1 = -0.0625
-- k _ _ = 0
-- @
--
scaleWithKernel
    :: (Int, Int)              -- ^ Output width, height
    -> (Int -> Int -> Double)  -- ^ kernel
    -> Image PixelRGBA8        -- ^ Input image
    -> Image PixelRGBA8        -- ^ Output image
scaleWithKernel dim@(w,h) kf = scaleImpl (cutWithKernel dim ker)
  where
    ker :: Array2D
    ker = dct1N [0, 1] $ array kb [ (i, kf x y) | i@(y, x) <- range kb ]
    kb = ((0, 0), (h-1, w-1))

scaleImpl :: (Array2D -> Array2D) -> Image PixelRGBA8 -> Image PixelRGBA8
scaleImpl cutImpl img = fromChannels r' g' b' a'
  where
    r = channelR img
    g = channelG img
    b = channelB img
    a = channelA img

    transform ch = amap (k*) ch'
      where
        ch' = dct3N [1, 0] . cutImpl . dct2N [0, 1] $ ch
        k = imgNorm ch / imgNorm ch'

    r' = transform r
    g' = transform g
    b' = transform b
    a' = transform a

imgNorm :: Array2D -> Double
imgNorm ch = sqrt . (/n) . sum . fmap sq . elems $ ch
  where
    sq x = x * x
    n = fromIntegral $ size ch

cut :: (Int, Int) -> Array2D -> Array2D
cut (w, h) img = array b [ (i, pick i) | i <- range b ]
  where
    b      = ((0,0), (h-1, w-1))
    b'     = bounds img
    pick i | inRange b' i = img ! i
           | otherwise    = 0

cutWithKernel :: (Int, Int) -> Array2D -> Array2D -> Array2D
cutWithKernel (w, h) k img = array b [ (i, pick i) | i <- range b ]
  where
    b      = ((0,0), (h-1, w-1))
    b'     = bounds img
    pick i | inRange b' i = k' i * (img ! i)
           | otherwise    = 0

    kb = bounds k
    k' i | inRange kb i = k ! i
         | otherwise    = 0

pixelR, pixelG, pixelB, pixelA :: PixelRGBA8 -> Word8
pixelR (PixelRGBA8 r _ _ _) = r
pixelG (PixelRGBA8 _ g _ _) = g
pixelB (PixelRGBA8 _ _ b _) = b
pixelA (PixelRGBA8 _ _ _ a) = a

extractChannel :: (PixelRGBA8 -> Word8) -> Image PixelRGBA8 -> Array2D
extractChannel f img@(Image w h _)
    = listArray ((0, 0), (h - 1, w - 1))
    . map (fromInteger . toInteger . f)
    . toListOf imagePixels
    $ img

channelR, channelG, channelB, channelA :: Image PixelRGBA8 -> Array2D
channelR = extractChannel pixelR
channelG = extractChannel pixelG
channelB = extractChannel pixelB
channelA = extractChannel pixelA

fromChannels
    :: Array2D
    -> Array2D
    -> Array2D
    -> Array2D
    -> Image PixelRGBA8
fromChannels r g b a = generateImage f w h
  where
    f x y = PixelRGBA8 (f' r) (f' g) (f' b) (f' a)
      where
        i = (y, x)
        f' c = truncate (limit $ c ! i)
    (_, (h', w')) = bounds r
    w = w' + 1
    h = h' + 1

limit :: Double -> Double
limit x | x < 0     = 0
        | x > 255   = 255
        | otherwise = x

-- From 'Control.Lens.Lens' from 'lens' package
toListOf :: Traversal s s a a -> s -> [a]
toListOf l = foldrOf l (:) []
{-# INLINE toListOf #-}

foldrOf :: Traversal s s a a -> (a -> r -> r) -> r -> s -> r
foldrOf l f z = flip appEndo z . foldMapOf l (Endo #. f)
{-# INLINE foldrOf #-}

foldMapOf :: Monoid r => Traversal s s a a -> (a -> r) -> s -> r
foldMapOf l f = getConst #. l (Const #. f)
{-# INLINE foldMapOf #-}

(#.) :: Coercible c b => (b -> c) -> (a -> b) -> a -> c
(#.) _ = coerce (\x -> x :: b) :: forall a b. Coercible b a => a -> b
{-# INLINE (#.) #-}

--(.#) :: Coercible b a => (b -> c) -> (a -> b) -> a -> c
--(.#) pbc _ = coerce pbc


example/Example.hs
module Main (main) where

import Codec.Picture          (convertRGBA8, readImage, writePng)
import Codec.Picture.ScaleDCT (scale, scaleWithKernel)
import Data.Time              (diffUTCTime, getCurrentTime)

processImage
    :: String      -- ^ From
    -> (Int, Int)  -- ^ Target size
    -> String      -- ^ To
    -> Maybe (Int -> Int -> Double) -- ^ kernel
    -> IO ()
processImage src size dst mk = do
    let scale' = maybe (scale size) (scaleWithKernel size) mk
    start <- getCurrentTime
    eimg <- readImage src
    case eimg of
        Left err -> print err
        Right dimg -> do
            let img = convertRGBA8 dimg
            let ava = scale' img
            writePng dst ava
            end <- getCurrentTime
            putStrLn $ dst ++ ": " ++ show (end `diffUTCTime` start)


main :: IO ()
main = do
    processImage "cat.jpg" (96, 64) "cat-small.png" Nothing
    processImage "cat.jpg" (769, 512) "cat-large.png" Nothing
    processImage "phadej.png" (32, 32) "phadej-smaller.png" (Just k)
    processImage "phadej.png" (64, 64) "phadej-small.png" (Just k)
    processImage "phadej.png" (600, 600) "phadej-large.png" Nothing
  where
    k 0 0 = 2.75 -- make this larger or smaller to adjust amount of an effect.
    k 0 1 = -0.125
    k 1 0 = -0.125
    k 1 1 = -0.0625
    k _ _ = 0
